### PR TITLE
Small optimization of acyclic graph merge operation

### PR DESCRIPTION
**Pull request by ppedrot**

### CODE REVIEW

**Comment by ppedrot** [comment]
Bench pending at https://gitlab.com/coq/coq/-/jobs/743414307.

**Comment by ppedrot** [comment]
Not bad:
```
┌─────────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬───────────────────────────────┬───────────────────┐
│                             │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]     │    mem faults     │
│                             │                         │                                             │                                             │                               │                   │
│                package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD PDIFF   │  NEW  OLD PDIFF   │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                         coq │  530.14  544.04 -2.55 % │     1482369285559     1519362638946 -2.43 % │     1996712785693     2073780487794 -3.72 % │     591952     592132 -0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                    coq-corn │ 1742.42 1759.99 -1.00 % │     4853375354045     4901090968465 -0.97 % │     7359999494419     7512958017401 -2.04 % │     855372     855884 -0.06 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│       coq-mathcomp-fingroup │   47.80   48.16 -0.75 % │      132943533357      133283810291 -0.26 % │      183716414184      183817372352 -0.05 % │     486516     486436 +0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                 coq-bignums │   57.96   58.28 -0.55 % │      160871828932      161097000033 -0.14 % │      215359118887      215382849415 -0.01 % │     472288     472160 +0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                    coq-hott │  314.49  316.18 -0.53 % │      853598931035      856184856887 -0.30 % │     1372570105654     1376108272794 -0.26 % │     675724     635912 +6.26 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│      coq-mathcomp-character │  159.88  160.56 -0.42 % │      445139314885      445959527550 -0.18 % │      618045390278      619163379591 -0.18 % │     720380     736856 -2.24 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-compcert │  662.92  665.44 -0.38 % │     1849138160875     1853576383757 -0.24 % │     2555374857290     2556071001846 -0.03 % │    1131048    1131380 -0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│             coq-lambda-rust │ 1259.96 1264.42 -0.35 % │     3508265326902     3517248032756 -0.26 % │     4936491855680     4950694508287 -0.29 % │    1214136    1174184 +3.40 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│        coq-mathcomp-algebra │  139.72  140.16 -0.31 % │      389152025535      389624333777 -0.12 % │      502772783614      503763086378 -0.20 % │     564492     558704 +1.04 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│             coq-fiat-crypto │ 4938.16 4953.20 -0.30 % │    13717014066689    13755417595288 -0.28 % │    23606611220758    23632654209960 -0.11 % │    2424624    2396068 +1.19 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│       coq-mathcomp-solvable │  171.74  172.26 -0.30 % │      478424051403      479465823364 -0.22 % │      676220411982      677686640682 -0.22 % │     691712     691976 -0.04 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                   coq-color │  616.60  618.45 -0.30 % │     1721141215184     1725603323172 -0.26 % │     2186829582488     2190464653436 -0.17 % │    1456172    1455236 +0.06 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│            coq-math-classes │  189.49  190.02 -0.28 % │      529462431757      530307410663 -0.16 % │      698532098261      700343222677 -0.26 % │     505276     504812 +0.09 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│            coq-fiat-parsers │  643.96  645.76 -0.28 % │     1803467963282     1807143934049 -0.20 % │     2716214484445     2718174531231 -0.07 % │    2997828    2997992 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│              coq-verdi-raft │ 1258.12 1261.08 -0.23 % │     3501931149566     3510464844375 -0.24 % │     4892910582698     4898226110532 -0.11 % │     935108     934680 +0.05 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-rewriter │  609.82  611.24 -0.23 % │     1692586519304     1694677581791 -0.12 % │     2577880404500     2584062241491 -0.24 % │    1032780     993028 +4.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│       coq-engine-bench-lite │  382.10  382.50 -0.10 % │     1074238348482     1076775012769 -0.24 % │     1836062983405     1839688187142 -0.20 % │    3508608    3508608 +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│          coq-mathcomp-field │  290.43  290.72 -0.10 % │      808063805858      809725635153 -0.21 % │     1247886203669     1249347899508 -0.12 % │     825780     826028 -0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-bedrock2 │  219.29  219.48 -0.09 % │      607686419276      607384560423 +0.05 % │     1003121904799     1003112567494 +0.00 % │    1087260    1087368 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                   coq-verdi │  121.61  121.64 -0.02 % │      335417576627      334989120197 +0.13 % │      413539133517      413666828270 -0.03 % │     574352     574448 -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│      coq-mathcomp-ssreflect │   50.31   50.28 +0.06 % │      138194716732      138023924566 +0.12 % │      176285027548      176404648113 -0.07 % │     523816     521060 +0.53 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                 coq-unimath │ 4003.05 3999.13 +0.10 % │    11141243812394    11125572015758 +0.14 % │    21171621630031    21168282190652 +0.02 % │    1113336    1113564 -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│ coq-rewriter-perf-SuperFast │ 1061.30 1060.26 +0.10 % │     2952646466468     2951308578932 +0.05 % │     4628293369085     4629516290838 -0.03 % │    1070732    1070776 -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                 coq-coqutil │   59.52   59.45 +0.12 % │      164875469160      165416892480 -0.33 % │      204986942946      205021747220 -0.02 % │     528836     528920 -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│      coq-mathcomp-odd-order │ 1069.59 1068.24 +0.13 % │     2980686786839     2976993251094 +0.12 % │     4747502530122     4750207627179 -0.06 % │    1021148     991632 +2.98 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│               coq-fiat-core │  105.68  105.52 +0.15 % │      301560252314      301925520661 -0.12 % │      401139285175      401563852425 -0.11 % │     488040     487208 +0.17 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                  coq-geocoq │ 1473.75 1471.08 +0.18 % │     4109973436218     4106620683384 +0.08 % │     6222236539194     6222341824576 -0.00 % │    1184268    1147804 +3.18 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│  coq-performance-tests-lite │  546.67  545.66 +0.19 % │     1527024869815     1524131625589 +0.19 % │     2795117184163     2795282688618 -0.01 % │     747308     747372 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                   coq-flocq │  160.77  160.36 +0.26 % │      445887801389      444478541972 +0.32 % │      553033885120      553184218432 -0.03 % │     822476     822132 +0.04 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│               coq-fourcolor │ 2524.50 2517.08 +0.29 % │     7038287491057     7017919689659 +0.29 % │    11443727152496    11447358132445 -0.03 % │     774816     774888 -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│              coq-coquelicot │   70.24   70.00 +0.34 % │      191434034550      191249055610 +0.10 % │      245211786998      245314124205 -0.04 % │     584976     587276 -0.39 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼───────────────────┤
│                coq-coqprime │   81.28   80.94 +0.42 % │      223010291003      222178962552 +0.37 % │      310431631768      310507946202 -0.02 % │     768212     768480 -0.03 % │    0    0  +nan % │
└─────────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴───────────────────────────────┴───────────────────┘
```

File: lib/acyclicGraph.ml [hunk 1]
```diff
@@ -359,13 +359,16 @@ module Make (Point:Point) = struct
     let ltle =
       let fold acc n =
         let fold u strict acc =
-          if strict then PMap.add u strict acc
-          else if PMap.mem u acc then acc
-          else PMap.add u false acc
+          match PMap.find u acc with
+          | true -> acc
+          | false -> if strict then PMap.add u true acc else acc
+          | exception Not_found -> PMap.add u strict acc
```
Comment (SkySkimmer) [inline lib/acyclicGraph.ml:365]: How about
```suggestion
          PMap.update u
            (function
              | None -> Some strict
              | Some strict' -> Some (strict || strict'))
            acc
```
?

Comment (ppedrot) [reply lib/acyclicGraph.ml:365]: Your variant allocates more. I don't know whether it's observable in practice but I'd prefer to stick to an allocation-free code when possible.

**Review by SkySkimmer — APPROVED** [review]
ok then

**Comment by SkySkimmer** [comment]
@coqbot: merge now

**Pull request merged**